### ISSUE TITLE
[no-unused-vars] is not reported in a scenario.

### ISSUE THREAD

**Issue by anshulguleria** [eslint/eslint#6348]
I am using 
`eslint@2.8.0`
`babel-eslint@6.0.4`
`node v6.0.0`
`npm 3.8.6`

My rule is set to `"no-unused-vars": 1`

```javascript
function test(options) {
    options = options || {};
    // Shouldn't this should give no-unused-vars error here?
}

test({});
```

In above scenario since options object is not used thus IMHO this should raise `no-unused-vars` error.

I tested this on online demo of eslint also.

**Comment by ilyavolodin** [comment]
Hmm.. This is a edge case. Technically an object is marked as used once it's value is assigned to something. In this case, however, it's assigned to itself, and rule is not smart enough to notice this. Covering all possible edge cases is not really possible, but I think we should try to cover this very standard pattern. @eslint/eslint-team thoughts?

**Comment by vitorbal** [comment]
I think this is a common enough use case that would be nice to have it covered.

Would we check if the object is assigned to itself *and* it is a function argument?
Or are you thinking of covering it in a more broad way, as in, this would also trigger an error after this enhancement:

```javascript
var foo = 1;
foo = foo + 1;
```

**Comment by ilyavolodin** [comment]
@vitorbal I think a good check is to see if the same variable is on both left and right side of the assignment expression. In that case, it should not be considered as "use".

**Comment by BYK** [comment]
I'm :+1: and am willing to champion.

**Comment by Slayer95** [comment]
imo this exception should only cover assignment if the RHS is a combination of LHS, a literal, and shortcircuiting operators.

Note that the snippet

```
function test(num) {
    num = num + 1;
}

test({
    toString() {
        console.log("called");
        return "1";
    }
});
```

<em>does</em> use the argument

**Comment by platinumazure** [comment]
I like @Slayer95's proposal. In that case there is a clear notion of potential non-use, which perhaps should be considered by the rule.

```js
myVar = myVar || someDefault();    // Potentially unused/not really reassigned, if myVar is truthy
myVar = myVar && process(myVar);   // Potentially unused/not really reassigned, if myVar is falsy

myVar = myVar + 1;                 // Clearly reassigned/used
```

**Comment by Slayer95** [comment]
Not sure about that ``&&`` example, for two reasons.

- If ``myVar`` is the LHS of ``&&``. then it can be regarded as a check.
On the other hand, for ``myVar = guard && myVar`` ``myVar`` can only be unused.

- If ``process`` has side effects depending on the argument, then ``myVar`` is definitely used, e.g.

``` js
function log(act) {
  logStream.write(act + '\n);
  notifyAction(act);
}

function process(act) {
  act = act && log(act);
}

process('account created');
```

Something interesting about these examples is that, although the variable is most certainly used, the assignment is pointless. But that shouldn't be a concern for this rule.

**Comment by Slayer95** [comment]
To summarize, ``x`` would not be flagged as consumed if in the following expressions:

`` x = a && x``
`` x = x || a``

Where ``a`` is not a function call, and is not an expression consuming ``x``.
I guess that would allow for a recursive implementation? But just dealing with the depth zero case should be enough: ``a`` being either a primitive / literal, or a variable.

**Comment by platinumazure** [comment]
The asymmetry of why RHS/LHS matters (and matters differently for `&&` vs `||`) is not clear to me.

**Comment by mysticatea** [comment]
I think this is reasonable and I can champion. :+1:

If somebody have the concern about side-effects, this is proper for an option.
Maybe `{ignoreInsideSelfAssign: true/false}`?

**Comment by ilyavolodin** [comment]
What about a case like this:

```js
function test(a) {
  a = a ? a : 0;
}
```

It's the same as shortcircuiting, but it's a more explicit format.

**Comment by platinumazure** [comment]
Hmm... I've reconsidered. I think it might be sufficient to provide an option which simply says, if a variable is on LHS and RHS of an assignment, don't assume it is used.

```js
function myFunc(a) {
    a = a + 1;
    // But wait, I'm still not really using a beyond the assignment,
    // so the assignment was actually pretty useless.
}
```

Also, I'm wondering about the compound-assignment operators. Do they count as usage?

```js
function myOtherFunc(a) {
    a += 1;  // Is this a "use" per no-unused-params?
}
```

If the compound-assignment doesn't count as a use, I fail to see how unwrapped assign-after-operator (`a = a + 1`) should count as a use either.

**Comment by Slayer95** [comment]
``a = a + 1`` could count as a use in as much this calls ``Number(a)``, which can have side effects.
However, as @mysticatea  says, that's better as a separate option.

I have also come to the opinion that a simple check for presence in both LHS and RHS is the best path for this for this issue.

**Comment by nzakas** [comment]
👍 for @platinumazure's suggestion

**Comment by mysticatea** [comment]
There seem to be enough :+1:s for `accepted`.
Remains... does it need an option or not?

**Comment by mikesherov** [comment]
:+1: for @platinumazure's suggestion, not optional. That is, until `a` is used for something other than self assignment, consider it unused.
